src/support/allocators/bulk_pool.h

```diff
@@ -0,0 +1,344 @@
+    size_t m_chunk_size{0};
+
+    //! Number of elements to allocate in bulk. Doubles each time, until m_max_num_allocs is reached.
+    size_t m_num_chunks_alloc{NUM_CHUNKS_ALLOC_MIN};
+};
+
+/// Allocator that's usable for node-based containers like std::unorderd_map or std::list. It requires a Pool
+/// which will be used for allocations of n==1. The pool's memory is only actually freed when the pool's
+/// destructor is called, otherwise previously allocated memory will be put back into the pool for further use.
+///
+/// Be aware that this allocator assumes that the containers will call allocate(1) to allocate nodes, otherwise
+/// the pool won't be used. Also, it assumes that the *first* call to allocate(1) is done with the actual node
+/// size. Only subsequent allocate(1) calls with the same sizeof(T) will make use of the allocator.
+template <typename T>
+class Allocator
+{
+    template <typename U>
+    friend class Allocator;
+
+    template <typename X, typename Y>
+    friend bool operator==(const Allocator<X>& a, const Allocator<Y>& b) noexcept;
+
+public:
+    using value_type = T;
+
+    //! Note: when two containers a and b have different pools and a=b is called, we replace a's allocator with b's.
+    using propagate_on_container_copy_assignment = std::true_type;
+    using propagate_on_container_move_assignment = std::true_type;
+    using propagate_on_container_swap = std::true_type;
+    using pointer = T*;
+    using const_pointer = const T*;
+    using reference = T&;
+    using const_reference = const T&;
+    using size_type = size_t;
+    using difference_type = std::ptrdiff_t;
+
+    template <typename U>
+    struct rebind {
+        using other = Allocator<U>;
+    };
+
+    explicit Allocator(Pool* pool) noexcept
+        : m_pool(pool)
+    {
+    }
+
+    /// Conversion constructor, all Allocators use the same pool.
+    template <typename U>
+    Allocator(const Allocator<U>& other) noexcept
+        : m_pool(other.m_pool)
+    {
+    }
+
+    /// Allocates n entries. When n==1, the pool is used. The pool might return nullptr if sizeof(T)
+    /// does not match the pool's chunk size. In that case, we fall back to new().
+    /// This method should be kept short so it's easily inlineable.
+    T* allocate(size_t n)
+    {
+        if (n == 1) {
+            auto r = m_pool->Allocate<T>();
+            if (r != nullptr) {
+                return r;
+            }
+        }
+        return reinterpret_cast<T*>(::operator new(n * sizeof(T)));
+    }
+
+    /// If n==1, we might have gotten the object from the pool. This is not the case when sizeof(T) does
+    /// not match the pool's chunk size. In that case, we fall back to delete().
+    void deallocate(T* p, std::size_t n)
+    {
+        if (n == 1 && m_pool->Deallocate(p)) {
+            return;
+        }
+        ::operator delete(p);
+    }
+
+    /// Calls p->~U(). Only required for g++4.8
```

Comment:
More information on this? A quick web search didn't turn up anything for me. Could say more in comment.

Reply:
It seems only g++4.8 requires that this method is present, otherwise I get a compile error. It's deprecated in C++17 and removed in C++20. I'll add a comment for that.